happi/audit.py

```diff
@@ -0,0 +1,423 @@
+"""
+from abc import ABCMeta, abstractmethod
+from configparser import ConfigParser
+import logging
+import os
+import sys
+import subprocess
+import re
+from happi.client import Client
+from happi.loader import import_class, fill_template
+from happi.containers import registry
+from enum import Enum
+
+logger = logging.getLogger(__name__)
+
+
+class ReportCode(Enum):
+    """
+    Class to define report codes
+    """
+    SUCCESS = 0
+    INVALID = 1
+    MISSING = 2
+    EXTRAS = 3
+    NO_CODE = 9
+
+
+class Command(object, metaclass=ABCMeta):
+    """
+    Bluprint for the command class
+    This will be useful if all the commands are inheriting from it
+    """
+
+    def __init__(self, name, summary):
+        self.name = name
+        self.summary = summary
+
+    @abstractmethod
+    def add_args(self, parser):
+        """
+        Adds arguments to a parser
+
+        Parameters
+        -----------
+        parser: ArgumentParser
+            Parser to add arguments to
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def run(self, args):
+        """
+        Parses the arguments given to the command.
+        And hanles the logic for this command
+
+        Parameters
+        -----------
+            args: Namespace
+        """
+        raise NotImplementedError
+
+
+class Audit(Command):
+    """
+    Audits the database for valid entries.
+    """
+    def __init__(self):
+        self.name = 'audit'
+        self.help = "Inspect a database's entries"
+        self._all_devices = set()
+        self._all_items = []
+        # 9 - NO_CODE
+        self.report_code = ReportCode(9)
+
+    def add_args(self, cmd_parser):
+        cmd_parser.add_argument(
+            "--file", help='Path to the json file (database) to be audited'
+        )
+        cmd_parser.add_argument(
+            "--extras", action='store_true', default=False,
+            help='Check specifically for extra attributes'
+        )
+
+    def run(self, args):
+        """
+        Validate the database passed in with --file option
+        """
+        if args.file is not None:
+            if self.validate_file(args.file):
+                logger.info('Using database file at %s ', args.file)
+                if args.extras:
+                    self.check_extra_attributes(args.file)
+                else:
+                    self.parse_database(args.file)
+            else:
+                logger.error('Probably provided a wrong path or filename')
+                return
+        else:
+            """
+            Validate the database defined in happi.cfg file
+            """
+            path_to_config = Client.find_config()
+
+            if path_to_config:
+                logger.info('Using Client cfg path %s ', path_to_config)
+                config = ConfigParser()
+                config.read(path_to_config)
+                try:
+                    database_path = config.get('DEFAULT', 'path')
```

Comment:
I'm really surprised we don't have a centralized code path for going from config to database file

Reply:
We do have this handled in the Client, `from_config(cls, cfg=None)` but it just does extra stuff that I don't need - but maybe I overcomplicated this here, I might just be able to use the `from_config` method, or if not, I could pull the config to database file portion out and place it in its own little function...

Reply:
This should be refactored around `from_config`, as you noted. I don't see a good reason to duplicate its work in a slightly different way?

Reply:
Yeah I see. I moved the `parsing configuration file` logic from `from_config`  into a separate `static method` in the `client` - in 75cbcd7. I am not sure if that was the right approach - but all I need is the `file_path` for the database, and `from_config` uses that but does not return me the path - what do you think?